src/rpc/services/SpectateService.ts

```diff
@@ -0,0 +1,531 @@
+                channelId
+            )
+        );
+        return currentOnChainSnapshot;
+    }
+
+    /**
+     * Run the payload in locally to verify it
+     */
+    private async verifyPayload(
+        channelId: ChannelId,
+        onChainSnapshot: StateSnapshot,
+        snapshotPayload: SnapshotPayload
+    ): Promise<{ isValid: boolean; error?: string; provenState?: any }> {
+        try {
+            const stateManager = this.mainRpcService.p2pManager.stateManager;
+
+            // Start with on-chain snapshot and traverse dispute windows until we reach the latest fork
+            let currentForkId = onChainSnapshot.forkId;
+            let isDisputed =
+                await stateManager.stateChannelManagerContract.isForkDisputed(
+                    channelId,
+                    currentForkId
+                );
+
+            while (isDisputed) {
+                // Check if reduced result already exists on-chain
+                const existingReducedResult =
+                    await stateManager.stateChannelManagerContract.getReducedResult(
+                        channelId,
+                        currentForkId
+                    );
+                if (existingReducedResult[0]) {
+                    // Traverse to the reduced fork
+                    currentForkId = existingReducedResult[0];
+                    isDisputed =
+                        await stateManager.stateChannelManagerContract.isForkDisputed(
+                            channelId,
+                            currentForkId
+                        );
+                    continue;
+                }
+
+                // No existing reduced result, use the disputes provided by participant
+                if (
+                    snapshotPayload.disputeWindows &&
+                    snapshotPayload.disputeWindows.length > 0
+                ) {
+                    // Pop the next dispute window
+                    const currentWindowDisputes =
+                        snapshotPayload.disputeWindows.shift()!;
+
+                    const creationTimestamp =
+                        await stateManager.stateChannelManagerContract.getDisputeWindowCreationTimestamp(
+                            channelId,
+                            currentForkId
+                        );
+
+                    // Reduce only this window
+                    const reducedOutput =
+                        await stateManager.stateChannelManagerContract.reduceProxyView(
+                            currentWindowDisputes,
+                            creationTimestamp
+                        );
+                    console.log(
+                        `Computed reduced output: forkGenesisTimestamp=${reducedOutput.forkGenesisTimestamp}`
+                    );
+
+                    // Move to next fork
+                    const nextForkId =
+                        reducedOutput.latestBlock.stateSnapshotHash;
```

Comment:
so the forkId is not the hash of the `StateSnapshot` but the hash of `SnapshotData`, but that doesn't change much here - you have to run the entire pipeline `reduceOutputToSnapshotData` which you can see happens in `reduceAndFinalize` - the `winningForkId` is your next fork